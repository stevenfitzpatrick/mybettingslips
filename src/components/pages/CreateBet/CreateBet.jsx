import React from 'react';
import { arrayOf, object, shape, string } from 'prop-types';
import { Button, Dropdown, Input, Modal } from '@sfitzpatrick/fitzy';
import { compose, graphql } from 'react-apollo';
import { Formik } from 'formik';
import { Mutation } from 'react-apollo';
import { number, object as yupObject, string as yupString } from 'yup';

import withAuth from '../../handlers/withAuth';
import {
  CreateBet as CreateBetMutation,
  GetBetOptions,
  GetUserBets
} from '../../../client/';

import { results } from '../../../store/bet.constants';
import { initialState } from './CreateBet.state';

const schema = yupObject().shape({
  stake: number('Must be integer').required('Stake is required'),
  odds: number('Must be integer').required('Odds is required'),
  typeId: yupString().required('Bet Type is required'),
  sportId: yupString().required('Sport Type is required'),
  awayId: yupString().required('Away Team is required'),
  homeId: yupString().required('Home Team is required')
});

export function CreateBet({
  allBetTypes,
  allSportTypes,
  allTeams,
  id,
  history
}) {
  function closeModal() {
    history.goBack();
  }

  return (
    <Modal onClose={closeModal}>
      <Mutation
        mutation={CreateBetMutation}
        update={(cache, { data: { createBet } }) => {
          const data = cache.readQuery({
            query: GetUserBets,
            variables: {
              first: 10,
              skip: 0
            }
          });

          data.user.bets = [createBet, ...data.user.bets];
          cache.writeQuery({
            query: GetUserBets,
            variables: {
              first: 10,
              skip: 0
            },
            data
          });
        }}
      >
        {createBet => (
          <Formik
            initialValues={initialState}
            onSubmit={async (
              { stake, odds, awayId, homeId, sportId, typeId },
              { setSubmitting, setErrors }
            ) => {
              try {
                await createBet({
                  variables: {
                    stake,
                    odds,
                    result: results.OPEN,
                    typeId,
                    awayId,
                    homeId,
                    sportId,
                    userId: id
                  }
                });
                setSubmitting(false);
                closeModal();
              } catch ({ message }) {
                setErrors({ message });
                setSubmitting(false);
              }
            }}
            render={({
              values,
              errors,
              handleChange,
              handleBlur,
              setFieldTouched,
              handleSubmit,
              setFieldValue,
              isSubmitting
            }) => (
              <form name="createBetForm" noValidate onSubmit={handleSubmit}>
                <Modal.Header>Add Bet</Modal.Header>
                <Modal.Body>
                  {errors.message && <div>{errors.message}</div>}
                  <Input
                    autoFocus
                    name="stake"
                    onBlur={handleBlur}
                    onChange={handleChange}
                    type="number"
                    value={values.stake}
                  />
                  <Input
                    name="odds"
                    onBlur={handleBlur}
                    onChange={handleChange}
                    type="number"
                    value={values.odds}
                  />
                  <Dropdown
                    items={allBetTypes}
                    itemToString={item => (item ? item.name : '')}
                    onBlur={() => setFieldTouched('typeId')}
                    onChange={i => setFieldValue('typeId', i.id)}
                    placeholder="Select Bet Type"
                  />
                  <Dropdown
                    items={allSportTypes}
                    itemToString={item => (item ? item.name : '')}
                    onBlur={() => setFieldTouched('sportId')}
                    onChange={i => setFieldValue('sportId', i.id)}
                    placeholder="Select Sport Type"
                  />
                  <Dropdown
                    items={allTeams}
                    itemToString={item => (item ? item.name : '')}
                    onBlur={() => setFieldTouched('homeId')}
                    onChange={i => setFieldValue('homeId', i.id)}
                    placeholder="Select Home Team"
                  />
                  <Dropdown
                    items={allTeams}
                    itemToString={item => (item ? item.name : '')}
                    onBlur={() => setFieldTouched('awayId')}
                    onChange={i => setFieldValue('awayId', i.id)}
                    placeholder="Select Away Team"
                  />
                </Modal.Body>
                <Modal.Footer>
                  <Button onClick={closeModal} use="Secondary">
                    Cancel
                  </Button>
                  <Button
                    disabled={isSubmitting}
                    loading={isSubmitting}
                    type="submit"
                  >
                    Submit
                  </Button>
                </Modal.Footer>
              </form>
            )}
            validationSchema={schema}
          />
        )}
      </Mutation>
    </Modal>
  );
}

CreateBet.propTypes = {
  allBetTypes: arrayOf(
    shape({
      id: string,
      name: string
    })
  ),
  allSportTypes: arrayOf(
    shape({
      id: string,
      name: string
    })
  ),
  allTeams: arrayOf(
    shape({
      id: string,
      name: string
    })
  ),
  history: object.isRequired,
  id: string.isRequired
};

CreateBet.defaultProps = {
  allBetTypes: [],
  allSportTypes: [],
  allTeams: []
};

export default compose(
  withAuth,
  graphql(GetBetOptions, {
    props: ({ getBetTypes: { allBetTypes, allSportTypes, allTeams } }) => {
      return { allBetTypes, allSportTypes, allTeams };
    },
    name: 'getBetTypes'
  })
)(CreateBet);
