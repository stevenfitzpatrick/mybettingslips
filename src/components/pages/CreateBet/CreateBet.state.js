export const results = {
  OPEN: 'Open',
  WIN: 'Win',
  LOSS: 'Loss',
  VOID: 'Void'
};

export const initialState = {
  stake: '',
  odds: '',
  result: results.OPEN,
  typeId: '',
  sportId: '',
  awayId: '',
  homeId: ''
};
