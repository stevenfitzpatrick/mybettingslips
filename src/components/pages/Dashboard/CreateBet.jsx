import PropTypes from 'prop-types';
import React from 'react';
import { compose, graphql } from 'react-apollo';
import { Dropdown, Input } from '@sfitzpatrick/fitzy';
import { Formik } from 'formik';
import { number, object, string } from 'yup';

import { BetTypesQuery, CreateBetMutation } from '../../../client/bets';
import { FormAlert } from '../../common';

const results = {
  OPEN: 'Open',
  WIN: 'Win',
  LOSS: 'Loss',
  VOID: 'Void'
};

const schema = object().shape({
  stake: number('Must be integer').required('Stake is required'),
  odds: number('Must be integer').required('Odds is required'),
  typeId: string().required('Type is required')
});

const propTypes = {
  createBet: PropTypes.func.isRequired,
  allBetTypes: PropTypes.array
};

const defaultProps = {
  allBetTypes: []
};

export function CreateBet({ createBet, allBetTypes }) {
  return (
    <div>
      <Formik
        initialValues={{
          stake: '',
          odds: '',
          result: results.OPEN,
          typeId: ''
        }}
        validationSchema={schema}
        onSubmit={async (inputs, { setSubmitting, setErrors }) => {
          try {
            await createBet({
              variables: {
                ...inputs
              }
            });
            setSubmitting(false);
          } catch ({ message }) {
            setErrors({ message });
            setSubmitting(false);
          }
        }}
        render={({
          values,
          errors,
          handleChange,
          handleBlur,
          setFieldTouched,
          handleSubmit,
          setFieldValue,
          isSubmitting
        }) => (
          <form onSubmit={handleSubmit} noValidate name="createBetForm">
            {errors.message && <FormAlert>{errors.message}</FormAlert>}
            <Input
              name="stake"
              type="number"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.stake}
              autoFocus
            />
            <Input
              name="odds"
              type="number"
              onChange={handleChange}
              onBlur={handleBlur}
              value={values.odds}
            />
            <Dropdown
              onChange={i => setFieldValue('typeId', i.id)}
              onBlur={() => setFieldTouched('typeId')}
              items={allBetTypes}
              placeholder="Select Sport Type"
            />
            <button>Cancel</button>
            <button type="submit" disabled={isSubmitting}>
              Submit
            </button>
          </form>
        )}
      />
    </div>
  );
}

CreateBet.propTypes = propTypes;
CreateBet.defaultProps = defaultProps;

export default compose(
  graphql(CreateBetMutation, {
    options({ id }) {
      return { variables: { userId: id } };
    },
    name: 'createBet'
  }),
  graphql(BetTypesQuery, {
    props: ({ getBetTypes: { allBetTypes } }) => {
      return { allBetTypes };
    },
    name: 'getBetTypes'
  })
)(CreateBet);
