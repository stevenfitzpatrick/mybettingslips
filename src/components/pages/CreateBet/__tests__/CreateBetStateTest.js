import { initialState, results } from '../CreateBet.state';

describe('CreateBet State', () => {
  test('should contain all correct Result options', () => {
    expect(results).toMatchInlineSnapshot(`
Object {
  "LOSS": "Loss",
  "OPEN": "Open",
  "VOID": "Void",
  "WIN": "Win",
}
`);
  });

  test('should have correct initial state for creating a bet', () => {
    expect(initialState).toMatchInlineSnapshot(`
Object {
  "awayId": "",
  "homeId": "",
  "odds": "",
  "result": "Open",
  "sportId": "",
  "stake": "",
  "typeId": "",
}
`);
  });
});
