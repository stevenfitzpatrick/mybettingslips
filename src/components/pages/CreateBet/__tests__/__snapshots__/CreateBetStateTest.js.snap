// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateBet State should contain all correct Result options 1`] = `
Object {
  "LOSS": "Loss",
  "OPEN": "Open",
  "VOID": "Void",
  "WIN": "Win",
}
`;
